single_kernel_mongo/events/backups.py

```diff
@@ -0,0 +1,252 @@
+    from single_kernel_mongo.managers.mongodb_operator import MongoDBOperator
+
+
+logger = logging.getLogger(__name__)
+
+INVALID_S3_INTEGRATION_STATUS = (
+    "Relation to s3-integrator is not supported, config role must be config-server."
+)
+
+
+class BackupEventsHandler(Object):
+    """Event Handler for managing backups and S3 integration."""
+
+    def __init__(self, dependent: MongoDBOperator):
+        super().__init__(parent=dependent, key="backup")
+        self.dependent = dependent
+        self.manager = self.dependent.backup_manager
+        self.charm: AbstractMongoCharm = dependent.charm
+        self.relation_name = ExternalRequirerRelations.S3_CREDENTIALS
+        self.s3_client = S3Requirer(self.charm, self.relation_name.value)
+
+        self.framework.observe(
+            self.charm.on[self.relation_name.value].relation_joined,
+            self._on_s3_relation_joined,
+        )
+        self.framework.observe(
+            self.s3_client.on.credentials_changed, self._on_s3_credential_changed
+        )
+        self.framework.observe(self.charm.on.create_backup_action, self._on_create_backup_action)
+        self.framework.observe(self.charm.on.list_backups_action, self._on_list_backups_action)
+        self.framework.observe(self.charm.on.restore_action, self._on_restore_action)
+
+    def _on_s3_relation_joined(self, event: RelationJoinedEvent) -> None:
+        """Checks for valid integration for s3-integrations."""
+        if self.dependent.state.upgrade_in_progress:
+            logger.warning(
+                "Adding s3-relations is not supported during an upgrade. The charm may be in a broken, unrecoverable state."
+            )
+            event.defer()
+            return
+        if not self.manager.is_valid_s3_integration():
+            logger.info(
+                "Shard does not support S3 relations. Please relate s3-integrator to config-server only."
+            )
+            self.charm.status_manager.to_blocked(INVALID_S3_INTEGRATION_STATUS)
+
+    def _on_s3_credential_changed(self, event: CredentialsChangedEvent):
+        action = "configure-pbm"
+        if self.dependent.state.upgrade_in_progress:
+            logger.warning(
+                "Changing s3-credentials is not supported during an upgrade. The charm may be in a broken, unrecoverable state."
+            )
+            event.defer()
+            return
+        if not self.manager.is_valid_s3_integration():
+            logger.debug(
+                "Shard does not support s3 relations, please relate s3-integrator to config-server only."
+            )
+            self.charm.status_manager.to_blocked(INVALID_S3_INTEGRATION_STATUS)
+            return
+        if not self.manager.workload.active():
+            defer_event_with_info_log(
+                logger, event, action, "Set PBM configurations, pbm-agent service not found."
+            )
+            return
```

Comment:
Leaving this for later with the status management but you cornered why it's made that way for now.
I haven't taken decision on `healthy` vs other solutions because not all events have the same checks.
I think however if there are some changes to make, I'll try to do it better.
Please keep in mind also that I tried to keep as much as possible the logic of the former workflows and that there might be some improvements like this that could be done.

But have a look as well: what is healthy for a backup manager ? valid S3 integration ? pbm-agent running ? No backup / restore in error ? Is it valid if we have credentials or if we don't ?

There are some reorganizations that we can do (and porting this in a method to be reused is one of them) but using some strong terms as `healthy` might not be accurate enough and could mislead IMHO.

Reply:
sounds good ! I will resolve, but remember this when we come back to health and status